src/cmd/flux-jobs.py

```diff
@@ -71,49 +56,83 @@ def runtime_fsd(job, hyphenifzero):
-            state_single=statetostr(job, True),
-            name=job["name"],
-            ntasks=job["ntasks"],
-            nnodes=job.get("nnodes", ""),
-            nnodes_hyphen=job.get("nnodes", "-"),
-            ranks=job.get("ranks", ""),
-            ranks_hyphen=job.get("ranks", "-"),
-            t_submit=job["t_submit"],
-            t_depend=job["t_depend"],
-            t_sched=job.get("t_sched", 0.0),
-            t_run=job.get("t_run", 0.0),
-            t_cleanup=job.get("t_cleanup", 0.0),
-            t_inactive=job.get("t_inactive", 0.0),
-            runtime=runtime(job, False),
-            runtime_fsd=runtime_fsd(job, False),
-            runtime_fsd_hyphen=runtime_fsd(job, True),
-            runtime_hms=runtime_hms(job),
-        )
-        print(s)
+class JobInfo:
+    """
+    JobInfo class: encapsulate job-info.list response in an object
+    that implements a getattr interface to job information with
+    memoization. Better for use with output formats since results
+    are only computed as-needed.
+    """
+
+    #  time attributes which may be missing in response
+    #  They should default to 0.0
+    #
+    t_attrs = ["t_sched", "t_run", "t_cleanup", "t_inactive"]
+
+    def __init__(self, info_resp):
+        self.info = info_resp.copy()
+        # Rename info.state to state_id until it is itself a string:
+        self.info["state_id"] = self.info.pop("state")
+
+    def __getattr__(self, attr):
+        #  First try to fetch attr from self.info, if that fails
+        #   then check if attr is a supported attribute and compute
+        #   it, memoizing the result in self.info for next access.
+        try:
+            return self.info[attr]
+        except KeyError:
+            pass
+
+        #  Computed attributes:
+        if attr in self.t_attrs:
+            #  If a legal time attribute is missing (see t_attrs above),
+            #   then default to 0.0
+            self.info[attr] = 0.0
+        elif attr == "runtime":
+            self.info[attr] = self.get_runtime()
+        elif attr == "runtime_fsd" or attr == "runtime_fsd_hyphen":
+            self.info[attr] = fsd(self.runtime, attr.endswith("_hyphen"))
+        elif attr == "runtime_hms":
+            t = self.get_runtime(True)
+            self.info[attr] = str(timedelta(seconds=t))
+        elif attr == "state":
+            self.info[attr] = statetostr(self.state_id)
+        elif attr == "state_single":
+            self.info[attr] = statetostr(self.state_id, True)
+        elif attr == "username" or attr == "user":
+            self.info[attr] = get_username(self.userid)
+        elif attr == "nnodes" or attr == "ranks":
+            self.info[attr] = self.info.get(attr, "")
+        elif attr == "nnodes_hyphen" or attr == "ranks_hyphen":
+            self.info[attr] = self.info.get(attr[:-7]) or "-"
+        else:
+            raise AttributeError("No jobinfo attribute " + attr)
+
+        return self.info[attr]
```

Comment:
When I gave this a glance last week, this is the only piece of the PR that gave me pause.  Normally, I would say that these should be implemented as methods to the object with the [`@property` decorator](https://docs.python.org/3/library/functions.html?highlight=property#property).   I even wrote up a whole mini example using it, but I realized it would be 2-3x the code.  Ultimately, I figured amplifying this code 2-3x and spreading it across the class would only increase confusion and the likelihood for bugs.  So I think it's fine as is.  I would be interested to see what @trws thinks.

Reply:
Good advice. I haven't delved into Python decorators yet, so regardless of outcome I'd appreciate seeing the example you coded up.

Reply:
I unfortunately typed it directly into the github comment and then deleted it (:man_facepalming:).  It looked something like this:

```python
import time

class JobInfo:
  def __init__(self, info_resp):
    for key, value in info_resp.items():
      setattr(self, "_{}".format(key), value)

  def get_runtime(self, roundup=False):
    if hasattr(self, "_t_cleanup") and hasattr(self, "_t_run"):
        t = self._t_cleanup - self._t_run
        if roundup:
            t = round(t + 0.5)
    elif hasattr(self, "_t_run"):
        t = time.time() - self._t_run
        if roundup:
            t = round(t + 0.5)
    else:
        t = 0.0
    return t

  @property
  def runtime(self):
    if hasattr(self, '_runtime'):
      return self._runtime
    self._runtime = self.get_runtime()
    return self._runtime

a = JobInfo({'t_run': 100000})
print(a.runtime)
```

```
❯ python3 ./test-class.py
1581981179.1356843
```

In this case, because a method named `runtime` with `@property.setter` above it has not been provided, `runtime` is a read-only property/attribute.

If you wanted to be really clever, you could dynamically create the property getter methods for the contents of `info_resp` within the `__init__` method.  You could also probably add a decorator on top of the `@property` decorator that auto-generates the memoization stuff.  All that seems too clever by half though, and I fully expect [Kernighan's Law](https://github.com/dwmkerr/hacker-laws#kernighans-law) would bite us in the butt.

Reply:
Nevermind, someone has already done the hard work for us: https://pypi.org/project/memoized-property/

This looks much nicer IMO and the code to implement the decorator is really small.

```python
import time
from functools import wraps

def memoized_property(fget):
    attr_name = '_{0}'.format(fget.__name__)

    @wraps(fget)
    def fget_memoized(self):
        if not hasattr(self, attr_name):
            setattr(self, attr_name, fget(self))
        return getattr(self, attr_name)

    return property(fget_memoized)

class JobInfo:
  # snip                                                                                                                                    

  @memoized_property
  def runtime(self):
    return self.get_runtime()
```

Reply:
Oh, that's pretty nice. Would you suggest copying memoized_property definition to `flux.util`? (If that is how you get to reuse these decorators?)

Reply:
> Would you suggest copying memoized_property definition to flux.util? (If that is how you get to reuse these decorators?)

Yeah, assuming it's OK to have differently licensed code in the same file.  If that is not OK (I'm assuming it is at minimum confusing), we could always just copy the `memoized_property.py` into the `src/bindings/flux` directory and access it with `from flux.memoized_property import memoized_property`.

Reply:
> If you wanted to be really clever, you could dynamically create the property getter methods for the contents of info_resp within the __init__ method.

This seems to be beyond me since AFAICT properties are set on a class and not per object.

Do you know what happens when an attribute is set with `setattr()` on an object, *and* a `@property` method is defined with the same name? The `JobInfo` class currently returns default values for some attributes that aren't set in the `info_resp`, e.g. the list of `t_attrs` as well `nodes`, `ranks` and `nodes_hyphen` `ranks_hyphen`

It would be convenient if the `@property` methods could just assume they are only called when the attribute has not already been set.

Reply:
Also, this change might have to wait until the exorcism of Python 2.

`functools` doesn't seem to export `wrap` in Python 2.7.

Reply:
Ugh, typo sorry! `wraps` not `wrap`

Reply:
> It would be convenient if the @property methods could just assume they are only called when the attribute has not already been set.

No, sadly, if you have an `@property` method set then `setattr()` fails. So you need a `@property.set` method, which means the value is no longer read-only. :meh:

Reply:
> This seems to be beyond me since AFAICT properties are set on a class and not per object.

Yeah, you are totally right.  I totally missed this nuance with [descriptors](https://docs.python.org/3/reference/datamodel.html#implementing-descriptors):

> The following methods only apply when an instance of the class containing the method (a so-called descriptor class) appears in an owner class (the descriptor must be in either the owner’s class dictionary or in the class dictionary for one of its parents). In the examples below, “the attribute” refers to the attribute whose name is the key of the property in the owner class’ __dict__."

`property` relies on descriptors in order to work correctly (it overrides the `__get__` descriptor) so you can't add a property to an instance without adding it to the class.

Good call!

---

> Do you know what happens when an attribute is set with setattr() on an object, and a @property method is defined with the same name?

The `@property` stuff gets setup at object creation, and the `__init__` method runs after that is all already setup.  So if you call `setattr` in `__init__`  with the exact name as the property, it will raise an exception because you will be trying to set a value on an property without a setter (i.e., a read-only attribute).   Note: it would work if `@runtime.setter`, for example, was defined for the class, then you could `setattr(self, "runtime", value)` without issue, but we want read-only.

This isn't an issue if you stuff the values in the response into the class in the same way that `memoized_property` does (i.e., with a `_` prefix).  For example:

```python
import time

class JobInfo:
  def __init__(self, info_resp):
    for key, value in info_resp.items():
      setattr(self, "_{}".format(key), value)

  @memoized_property
  def runtime(self):
    return time.time() - self._t_run

a = JobInfo({'t_run': 100000, 'runtime': 42})
print(a.runtime)
```

```
→ python3 ./test-class.py
42
```

If you still want to support the case where values included in the response can be directly accessed as attributes, you can still include the first part of the `__getattr__` override you originally had:

```python
    def __getattr__(self, attr):
        try:
            return getattr(self, "_{}".format(attr))
        except KeyError:
            raise AttributeError
```

With that in the above example, `print(a.t_run)` then prints `100000`.

I hope that answers your question (I'm just pseudo-randomly smacking keys on my keyboard at this point in the night TBH).

Reply:
Your suggestion for `__getattr__` would be great, however it causes infinite recursion for a non-existent property (including `runtime` since `memoize_property` calls the `getattr` method for the class)

Avoiding recursion in `getattr` is why the original function proposed here stored the data in a "shadow" dict inside the object (`self.info`). E.g. this works:

```Python
class JobInfo:
  def __init__(self, info_resp):
    self.info = info_resp.copy()

  def __getattr__(self, attr):
    try:
      return self.info[attr]
    except KeyError:
      raise AttributeError

  @memoized_property
  def runtime(self):
    return time.time() - self.t_run

```

Though Python doesn't seem to like rethrowing a different exception as in the getattr above -- when I add `print(a.dummy)` I get:

```
Traceback (most recent call last):
  File "test-class.py", line 23, in __getattr__
    return self.info[attr]
KeyError: 'dummy'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test-class.py", line 37, in <module>
    print(a.dummy)
  File "test-class.py", line 25, in __getattr__
    raise AttributeError
AttributeError
```

Reply:
@SteVwonder: I've pushed a couple commits on top of my current work that converts `JobInfo` to use `@memoized_property` functions for the computed job values. I kept `self.info` as a shadow dict of values to avoid the infinite recursion caused by `__getattr__` calling `getattr()`.

Let me know what you think -- I don't think it is too bad where we ended up.

The result is a *wee* bit slower than the previous approach proposed, likely due to the extra function wrapping caused by `@memoized_property` (but it is almost in the noise, so nothing to worry about I think)

Reply:
> Your suggestion for `__getattr__` would be great, however it causes infinite recursion for a non-existent property (including `runtime` since `memoize_property` calls the `getattr` method for the class)

It would, but are there ever meant to be any attributes that begin with "_" otherwise? If not, perhaps:

```python
    def __getattr__(self, attr):
        if attr.startswith("_"):
            raise AttributeError
        try:
            return getattr(self, f"_{attr}")
        except KeyError:
            raise AttributeError
```

Reply:
Well, that seems reasonable and very clever! Will it work with `memoized_property` though? The decorator seems to call `getattr(self, attr_name)` to check for a memoized result, where `attr_name` is the `_` prefixed name of attr.

However, I am totally willing to admit I'm being dumb and missing something obvious.

Reply:
That's where using `__getattr__` is important rather than `__getattribute__`, the `__getattr__` method only gets called if the normal lookup has already failed, so the first time in it gets forwarded back to look up the `_` prefixed version, and if it finds it `__getattr__` doesn't get entered again.

Reply:
Ah ok! I got it working now, there was something strange in the way I was handling default values. I've pushed a commit that switches to your version of `__getattr__` as well as a redo on how class default values are handled. If everything looks ok, I'll squash incremental development.